### PR TITLE
Optimizations + target DNS injection

### PR DESCRIPTION
**Pull request by deputinizer**
Currently looking at pprof, more changes coming.

### CODE REVIEW

File: internal/stoppropaganda/customresolver/customgetip.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	addr := net.ParseIP(host)
+	if addr == nil {
+		ips, err := CustomLookupIP(host, helperIPBuf)
+		if err != nil {
+			return nil, err
+		}
+		for i := 0; i < len(ips); i++ {
+			ip := ips[i]
+			if ipv4 := ip.To4(); ipv4 == nil {
+				// Remove inplace trick
```
Comment (baylanger) [inline internal/stoppropaganda/customresolver/customgetip.go:36]: Thank you for putting some comments, trying to learn go ;)

Comment (deputinizer) [reply internal/stoppropaganda/customresolver/customgetip.go:36]: Yeah, that's a neat trick

**Comment by deputinizer** [comment]
<details>
  <summary>Still, dial is the main CPU eater</summary>

image

</details>

**Comment by deputinizer** [comment]
<details>
  <summary>[Spoiler] Memory usage from pprof</summary>
  
image

</details>

**Comment by deputinizer** [comment]
Why there was no `getRandomDomain()` in every DNS request?

**Comment by deputinizer** [comment]
Yeah, now it runs constant 100 Mbit/s :100:

**Comment by erkexzcx** [comment]
Good to be merged?

**Comment by erkexzcx** [comment]
> Why there was no `getRandomDomain()` in every DNS request?

Because it was CPU waste which does not bring any benefit. I did not want to use `russiagofuckyourself.ru` or something like that, since it would make very easy to filter out such requests if they want to check their logs. Creating a random domain once during startup would be sufficient IMO.

**Comment by deputinizer** [comment]
Works, so far it didn't panic :+1:

**Comment by deputinizer** [comment]
> Creating a random domain once during startup would be sufficient IMO.

But they can cache the 404 response and return instantly.

**Comment by deputinizer** [comment]
> Good to be merged?

Oh, let's also reuse the `fastrand.RNG`. Wait a minute

**Comment by erkexzcx** [comment]
> But they can cache the 404 response and return instantly.

ohhh that makes sense. Randomizing domains actually helps then.

**Comment by deputinizer** [comment]
All done.

**Comment by erkexzcx** [comment]
Let me do some tests and merge.

**Comment by erkexzcx** [comment]
P.S. Can you try to fix this? https://github.com/erkexzcx/stoppropaganda/issues/100

EDIT: Doesn't seem to appear with your PR code.

**Comment by erkexzcx** [comment]
Good job

**Comment by erkexzcx** [comment]
With this PR it even started to work on Android. Perfection...

**Pull request merged**